pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const clusterScopedNamespace = ""
+
+// placementResolution stores the selected objects and destinations for a single placement.
+// The mutex should be read locked before reading, and write locked before writing to any field.
+type placementResolution struct {
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
```

Comment:
If the keys in this map used _resource_ instead of _kind_ then there would be less converting between the two. For example, `func (resolver *placementResolver) GeneratePlacementDecision` would not need the ability to return an error.

Reply:
Assuming you are not referring to refactoring the util.Key struct to include resource name but are suggesting extending the value of this map to be {Key, resource}:
While this would save conversions between the two in GeneratePlacementDecision, the conversions would be paid in UpdateObject (or further up its stack). I'm not sure this is worth it.

If what you are suggesting is refactoring util.Key itself, I believe this can be done in a follow-up work if you don't deem it a necessity for current PR.

Reply:
I am not suggesting making Key hold redundant information.

My comment is an FYI, not a blocker. In fact, I would probably prefer the change to be a separate PR. The idea is that the pervasive focus on Kind rather than resource leads to more places in the code needing gvr/gvk mapping than if the pervasive focus were on resource.

Reply:
I completely agree. 

Just for information: the reasoning for not changing this (and some other choices in the code) is to stay as close as possible to the current implementation and its flow - whether good or bad - since we dont want to propose a full re-implementation of placement. When this PR goes through, this and other changes will be better backed.